implementations/rust/kex/src/xx.rs

```diff
@@ -106,31 +111,43 @@ impl KeyExchange for SymmetricState {
         &mut self,
         secret_handle: SecretKeyContext,
         public_key: PublicKey,
-    ) -> Result<Vec<u8>, VaultFailError> {
+    ) -> Result<Vec<SecretKeyContext>, VaultFailError> {
         let mut vault = self.vault.lock().unwrap();
+
+        let attributes_ck = SecretKeyAttributes {
+            xtype: SecretKeyType::Buffer(32),
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
+        let attributes_k = SecretKeyAttributes {
+            xtype: SecretKeyType::Aes256,
+            purpose: SecretPurposeType::KeyAgreement,
+            persistence: SecretPersistenceType::Ephemeral,
+        };
+
         vault.ec_diffie_hellman_hkdf_sha256(
             secret_handle,
             public_key,
-            self.state.ck.as_ref(),
-            SHA256_SIZE + AES256_KEYSIZE,
+            self.state.ck.unwrap(),
+            vec![attributes_ck, attributes_k],
         )
     }
 
     /// mix key step in Noise protocol
-    fn mix_key<B: AsRef<[u8]>>(&mut self, hash: B) -> Result<(), VaultFailError> {
-        let hash = hash.as_ref();
-        self.state.ck.copy_from_slice(&hash[..SHA256_SIZE]);
-        let attributes = SecretKeyAttributes {
-            xtype: SecretKeyType::Aes256,
-            persistence: SecretPersistenceType::Ephemeral,
-            purpose: SecretPurposeType::KeyAgreement,
-        };
-        let key = SecretKey::Aes256(*array_ref![hash, SHA256_SIZE, AES256_KEYSIZE]);
+    fn mix_key(&mut self, hkdf_output: Vec<SecretKeyContext>) -> Result<(), VaultFailError> {
         let mut vault = self.vault.lock().unwrap();
+
+        if self.state.ck.is_some() {
+            vault.secret_destroy(self.state.ck.unwrap())?;
+        }
+        self.state.ck = Some(hkdf_output[0]);
```

Comment:
prefer use of `.get(0)` on vec and to handle the `Option` returned, or do a length check before and handle the case where invalid.

Reply:
Agree, tnx